File: moqt/internal/message/announce_please.go
package message

import (
	"io"
)

/*
 *	ANNOUNCE_PLEASE Message {
 *	  Track Prefix (string),
 *	}
 */
type AnnouncePleaseMessage struct {
	TrackPrefix string
}

func (aim AnnouncePleaseMessage) Len() int {
	return StringLen(aim.TrackPrefix)
}

func (aim AnnouncePleaseMessage) Encode(w io.Writer) error {
	msgLen := aim.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteString(b, aim.TrackPrefix)

	_, err := w.Write(b)

	return err
}

func (aim *AnnouncePleaseMessage) Decode(src io.Reader) error {
	num, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	b := make([]byte, num)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	str, n, err := ReadString(b)
	if err != nil {
		return err
	}
	aim.TrackPrefix = str
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: interop/web/src/message/announce_please.ts
import { Writer, Reader } from "../io";
import { stringLen } from "../io/len";

export class AnnouncePleaseMessage {
    prefix: string;

    constructor(prefix: string) {
        this.prefix = prefix;
    }

    length(): number {
        return stringLen(this.prefix);
    }

    static async encode(writer: Writer, prefix: string): Promise<[AnnouncePleaseMessage?, Error?]> {
        const msg = new AnnouncePleaseMessage(prefix)
        writer.writeVarint(BigInt(msg.length()));
        writer.writeString(prefix);
        const err = await writer.flush();
        if (err) {
            return [undefined, err];
        }
        return [msg, undefined];
    }

    static async decode(reader: Reader): Promise<[AnnouncePleaseMessage?, Error?]> {
        const [len, err] = await reader.readVarint();
        if (err) {
            return [undefined, new Error("Failed to read length for AnnouncePlease")];
        }
        const [str, err2] = await reader.readString();
        if (err2) {
            return [undefined, new Error("Failed to read prefix for AnnouncePlease")];
        }

        return [new AnnouncePleaseMessage(str), undefined];
    }
}